HOUSNAV-47: 9.9.9.1 ui questions

nicholaslyon-dt opened the pull request:
[HOUSNAV-47](https://hous-bssb.atlassian.net/browse/HOUSNAV-47)

## Overview & Purpose
Add the logic necessary for getting through the questions associated with 9.9.9.1

## Summary of Changes
Refactored currentItemId to be in the NavigationStore and to use a getter and setter for extra logic tied to when it's set
Added the ability for MultiChoiceMultiple data to be an array or an object
Added logic to dynamically show possible answers
Grouped more navigation state changes together
Added logic to handle questions that just set a variable
Marked some questions as not requiring an answer and added functionality for that
Added logic to set a default answer for some questions that are not required
Updated back logic to handle questions in history that just set a variable
Added type checking utilities
Added next navigation logic types and updated the base function to be recursive for and and or types
Added missing types in walkthrough data
Cleaned up walkthrough data errors and added new properties

## Side Effects
Could effect the intro questions, but I have not been skipping them to be able to test new questions

## Testing
Answer intro and questions in 9.9.9.1
Move back and forth changing answers and compare against [test case 1](https://docs.google.com/document/d/1VToXTJ10rf__5g771tb5KhCvihJxZPouY5XkYqEnLgk/edit)

Commit:
HOUSNAV-47: Initial commit for 9.9.9.1 ui questions

Commit:
HOUSNAV-47: Small refactor of NavigationStore for clarity and organization

jbook24 commented on apps/web/utils/logic/showAnswer.ts:
```diff
@@ -0,0 +1,77 @@
+  getAnswerToCheckValue: AnswerToCheckValueFn,
+) => {
+  return possibleAnswers.filter((possibleAnswer) => {
+    if (!possibleAnswer.showAnswerIf) return true;
```
Could we do something like this? 
`if (possibleAnswer.showAnswerIf == null || possibleAnswer.showAnswerIf === true) return true;
`

nicholaslyon-dt replied:
Not quite because of a quirk with TypeScript and JSON. the `showAnswerIf` property can be `boolean | ShowAnswerIf[] | undefined`. But really, it can never be false. I just can't set it to `true | ShowAnswerIf[] | undefined` because of the type difference issues between TypeScript and JSON. I will update this to absorb the check below it though. LMK what you think.

jbook24 commented on apps/web/utils/logic/showAnswer.ts:
```diff
@@ -0,0 +1,77 @@
+    showAnswer = true;
+  }
+
+  return showAnswer;
```
Could we just ? 

```
// Return true if answerToCheck is greater than answerValue, otherwise false
return answerToCheck > answerValue;
```

nicholaslyon-dt replied:
Yes! Good call.

jbook24 commented on apps/web/utils/logic/variableItem.ts:
```diff
@@ -0,0 +1,110 @@
+  getAnswerToCheckValue: AnswerToCheckValueFn,
+): AnswerTypes => {
+  // check variableType
+  switch (variableToSet.variableType) {
```
Given we only have one case in our switch could we just? 
```

if (variableToSet.variableType === VariableToSetType.Object) {
  return ThisModule.getVariableItemValueObject(variableToSet, getAnswerToCheckValue);
}

```

nicholaslyon-dt replied:
Yeah, I had left this as is because there was a potential for another type, but at the moment there is not!

Commit:
HOUSNAV-47: logic for skipping questions + PR feedback + more tests

Commit:
HOUSNAV-47: adding more testing and errors for logic code plus formatting and linting fixes

Commit:
HOUSNAV-47: more testing and errors for logic code

wardm89 commented:
![Screenshot 2024-06-14 at 2 21 53 PM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/7059124/62d97188-4b41-412a-a979-3e0b044bceba)
I just got this error. So far, I'm trying to replicate it but idr my answers I did to get it :) But I do know it was after clicking `2 stories`.

Any thoughts on this?
I'm assuming it came through undefined.

Commit:
HOUSNAV-47: JSON logic fix

nicholaslyon-dt commented:
> ![Screenshot 2024-06-14 at 2 21 53 PM](https://private-user-images.githubusercontent.com/7059124/339867192-62d97188-4b41-412a-a979-3e0b044bceba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzOTI2NTYsIm5iZiI6MTcxODM5MjM1NiwicGF0aCI6Ii83MDU5MTI0LzMzOTg2NzE5Mi02MmQ5NzE4OC00YjQxLTQxMmEtYTk3OS0zZTBiMDQ0YmNlYmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTRUMTkxMjM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjQ2NWM2Zjc3NjE5NzY2MTNkNGI3YWM4ZWVkYzZhMTUzZDdjYzZlZTA2ZTJmYjc3ZmU0ZGE0NTU4YWYyMDgyZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.zKgfA60qTMIk8rIFCFdqff3le9G3Vusn9w3wHyv_YII) I just got this error. So far, I'm trying to replicate it but idr my answers I did to get it :) But I do know it was after clicking `2 stories`.
> 
> Any thoughts on this? I'm assuming it came through undefined.

Hrm I'm not sure. Possible Invalid Answers should only be called in 3 cases and they all happen in the intro questions. I've been running through the questions back and forth and can't get that to happen.

Commit:
HOUSNAV-47: Added console log error handling when handling forward navigation

wardm89 commented on apps/web/utils/logic/nextNavigation.ts:
```diff
@@ -38,106 +81,249 @@ export const getNextNavigationId = (
+    );
+  }
+
+  return isNotEqual;
```
```suggestion
  if (isString(answerToCheck)) {
    return answerToCheck !== answerValue;
  }

  if (answerToCheck === undefined) {
    return answerValue !== "undefined";
  }
  // if answerToCheck is not a string or undefined, throw an error
  throw new Error(
    `nextLogicTypeNotEqual: answerToCheck must be a string or undefined, got ${typeof answerToCheck}`
  );
```

wardm89 commented on apps/web/utils/logic/nextNavigation.ts:
```diff
@@ -38,106 +81,249 @@ export const getNextNavigationId = (
+    throw new Error(
+      `nextLogicTypeEqual: answerToCheck must be a string, array, or undefined, got ${typeof answerToCheck}`,
+    );
   }
```
Very similar recommendation to the other. We don't actually need all the else if statements and can simplify this.

Commit:
HOUSNAV-47: actually showing error screen when next navigation error occurs

wardm89 commented on apps/web/utils/logic/showAnswer.ts:
```diff
@@ -0,0 +1,91 @@
+  });
+};
+
+export const showAnswerTypeEquals = (
```
This can be shortened to remove the if else statements as well.

if(isArray)return...
if(isString)return...
throw error

wardm89 commented on apps/web/utils/logic/variableItem.ts:
```diff
@@ -0,0 +1,110 @@
+  throw new Error("Invalid variableType");
+};
+
+export const getVariableItemValueObject = (
```
I feel like this could be shortened a lot by deconstructing the variables and then not needing `valueHasBeenSet` since we can just `break` the loop after each `newObjectVariable[key] =` line.

wardm89 commented on apps/web/utils/logic/nextNavigation.ts:
```diff
@@ -1,31 +1,74 @@
-    const answerToCheck = answerState[navigationLogicItem.answerToCheck || ""];
+    const answerToCheck = getAnswerToCheckValue(
+      navigationLogicItem.answerToCheck,
+    );
```
What is the difference between this `answerToCheck` and the one that could be deconstructed from `navigationLogicItem`?

nicholaslyon-dt replied:
That's the value. Refactored across the board for clarity.

wardm89 commented on apps/web/stores/AnswerStore.ts:
```diff
@@ -2,28 +2,116 @@
+      // erase future state
+      this.eraseAnswersAfterCurrentItem();
+
+      // set answer value
+      this.answers[currentId] = newItemValue;
```
This goes for a lot of these comments. I feel they are redundant. I think you've done a good job naming functions/methods as well so things like this below is also redundant.
```
// add variable item to history
 addItemIdToHistory(currentId);
```

nicholaslyon-dt replied:
Refactored across the board! LMK what you think.

wardm89 reviewed (changes requested):
I think with some refactoring to deconstruct variables so lines don't wrap unnecessarily, simplifying some if/else logic and removing comment's it'll make this a lot more readable. 

As far as my "bug" I had before. I can't replicate it but with the new error handling on navigation that will enable us to replicate it if it happens again.

wardm89 commented:
One last comment. With all the errors being thrown, they aren't being handled anywhere currently or are they?

Commit:
HOUSNAV-47: PR feedback refactoring

Commit:
HOUSNAV-47: Clarify refactor for difference between answerToCheck and answerToCheckValue

Commit:
HOUSNAV-47: Refactored comments to be more about business logic and project knowledge

Commit:
HOUSNAV-47: error handling function addition

nicholaslyon-dt commented:
> One last comment. With all the errors being thrown, they aren't being handled anywhere currently or are they?

Good call. Most were caught by the handling we added for the forward navigation state, but there was one case that still needed caught. I added a function for handling and caught it in the other spot as well.

Commit:
HOUSNAV-47: remove linting fix to match other PR
